### PR TITLE
Only bind module.exports if no local definition exists

### PR DESCRIPTION
**Pull request by sandersn**
Note that this uses `lookupSymbolForNameWorker`, which is really abest-effort check since it only knows about symbols that it has already encountered.

As a side-effect, even when `module` is bound as part of a `module.exports` reference, it only declares it once instead of one declaration per reference.

Fixes user baseline regression in webpack as found in #25866.

### CODE REVIEW

**Comment by weswigham** [comment]
@sandersn can you also add a test with
```ts
function exec() {
    const module = new C(12);
    return () => {
        return module.exports; // should be fine because `module` is defined locally
    }
}
```

I don't think `lookupSymbolForNameWorker` is sufficient here.

**Comment by sandersn** [comment]
Nice test -- that exposed a crash that happens because we erroneously create a symbol with ModuleExports on it in a script file. Since script files, unlike module files, don't have a symbol, the checker is unable to create the synthetic type for `module: { exports: import("~/src/my/filename") }`.

The binder doesn't know ahead of time whether it is in a commonjs module or not -- that property is calculated lazily inside the binder, so I just added a check in the checker.

File: tests/baselines/reference/moduleExportAssignment6.types [hunk 1]
```diff
@@ -49,3 +49,23 @@ function exec() {
...
+
+    // (a trickier variant of what webpack does)
+    const module = new C(12);
+>module : C
+>new C(12) : C
+>C : typeof C
+>12 : 12
+
+    return () => {
+>() => {        return module.exports;    } : () => any
```
Comment (weswigham) [inline tests/baselines/reference/moduleExportAssignment6.types]: This return type is wrong. AFAIK, `module.exports` here should be a `number[]`.

**Comment by sandersn** [comment]
OK, I made the binder more conservative about creating a symbol for `module.exports`. See the commit notes for details.

**Pull request merged**